Send sqs message for speech to text create

peetucket opened the pull request:
## Why was this change made? 🤔

Fixes #1356 - send an SQS message to kick off speech to text (after files sent to S3)

## How was this change tested? 🤨

New specs

edsu commented on lib/robots/dor_repo/speech_to_text/stt_create.rb:
```diff
@@ -11,11 +11,35 @@ def initialize
+        def send_sqs_message
+          # Define the queue URL and message body
+          queue_url = Settings.aws.sqs_queue_url
+          message_body = {
```
The body of the message that speech-to-text is currently looking for is documented here:

https://github.com/sul-dlss/speech-to-text/blob/docker/README.md#the-job-file

peetucket replied:
So is the idea that we need to send one message per file that we want transcribed?  if so, then we can iterate over the druid folder in S3, find all of the files, and send multiple messages

though then...how would we know when all of the files in a given druid are done?

jmartin-sul replied:
oh, hmm, that's a good point.  maybe that's an argument for having the speech-to-text service do a directory or a list of files, instead of one file at a time.

alternatively, i can vaguely imagine some ways that the watcher could figure out whether a particular success/failure message for one file, possibly among many for a druid version, is the last one being processed, allowing the step to get marked completed if there were no errors in the batch:
* we could use WFS vars, with e.g. a hash of files to process by the wf for the druid version (hash initialized in this wf step), the processing status updated for each file as it's processed (set upon SQS message receipt by the [watcher daemon we've been discussing today](https://github.com/sul-dlss/common-accessioning/issues/1358)).  and then each time an update is received, the final success/error state of the step could be set?
* we could give common-accessioning its own data store to track that info, but otherwise use similar logic.

i'd probably vote against adding a database to common-accessioning, with no preference between the other two options (and interested in other better options if anyone can think of any).

adding extra load/responsibility to WFS seems not great, but this also seems along the lines of what WFS variables are meant for.  and it doesn't seem like it'd be adding that much load, if we only have a small handful of worker containers processing STT jobs.

keeping the speech-to-text worker container simple also seems desirable, but processing a list of files instead of one file isn't a terrible extra complication (sorry for the back and forth if we do go that route @edsu, i feel like we talked you into one file at a time when we all initially met to look at your work on https://github.com/sul-dlss/speech-to-text/pull/9).

jmartin-sul replied:
given the bare-bones nature of both our argo GUI and workflow-server-rails/lyber-core CLI tools for interacting with robots and workflow state, i'd guess that we shouldn't try to make individual media file extraction failures retriable, and that we should just let the whole batch for the druid version fail if one file fails, then fix the issue and retry the step if need be.  if we run into a bunch of obviously automatically retriable errors, maybe we can revisit and add retry handling for those things where ever is most appropriate?

jmartin-sul replied:
another thought, which could be an optimization: upon successful text generation for a given media file, the watcher process downloads the output for that media file.  then the speech-to-text job queuing code in this class could look to see if there is any output already, and only queue jobs for media files requiring STT that don't already have output.  that saves needless re-running and and thus GPU waste, without having to think about trickier retry capabilities.

jmartin-sul replied:
we'll might need to update https://github.com/sul-dlss/common-accessioning/issues/1358 with whatever comes out of this thread.

jmartin-sul replied:
oh, heh, i may have written all that for nothing: looking back again at the README in https://github.com/sul-dlss/speech-to-text/pull/9/, and at the `download_media` code in that PR for the first time (sort of reviewing these two in parallel, and occasionally flipping over to #1350)... it seems like it already handles a list with multiple files, since the `media` field in the job structure is an array.

i guess the "download all the successful extractions in a partially failed batch" even if it the whole batch doesn't succeed is still a useful thought, and i might've gotten there faster for all the likely useless writing above that 😆

jmartin-sul replied:
also, `SecureRandom.uuid` for the job ID?

peetucket replied:
Thanks for summarizing.

jmartin-sul commented on config/settings.yml:
```diff
@@ -58,4 +58,6 @@ aws:
-  base_s3_bucket: 'sul-speech-to-text-dev'
+  speech_to_text:
+    base_s3_bucket: 'sul-speech-to-text-dev' # default bucket for storing speech-to-text file
+    sqs_new_job_queue_url: 'https://sqs.us-west-2.amazonaws.com/queue_url' # queue to send new speech-to-text jobs
```
what about `sqs_done_queue_url`, to mirror the naming in https://github.com/sul-dlss/speech-to-text/pull/9?

jmartin-sul replied:
heh, i'd actually meant to say `sqs_todo_queue_url`, but thanks for adding the one i mentioned and doing the rename!

peetucket commented:
1. Added a UUID to the message body
2. Added the media files to the message.  

I wonder if instead of reaching out to the cocina object itself to find the list of SSTable files to add to the job, we should instead just look in the AWS bucket folder that should have been created in the previous workflow step and just add any files there.  Could be convinced of either approach, i think they both have pros/cons but probably doesn't matter too much one way or the other.

peetucket commented on lib/robots/dor_repo/speech_to_text/fetch_files.rb:
```diff
@@ -23,11 +23,7 @@ def aws_provider
-        def speech_to_text
-          @speech_to_text ||= Dor::TextExtraction::SpeechToText.new(cocina_object:, workflow_context: workflow.context)
+          Dor::TextExtraction::SpeechToText.new(cocina_object:, workflow_context: workflow.context).filenames_to_stt
         end
```
small refactor ... we don't really need this much indirection since the `filenames_to_stt` is only used once in the robot

peetucket commented on lib/robots/dor_repo/speech_to_text/stt_create.rb:
```diff
@@ -11,10 +11,51 @@ def initialize
+
+        def media
+          Dor::TextExtraction::SpeechToText.new(cocina_object:, workflow_context: workflow.context).filenames_to_stt
+        end
```
this will reach out to DSA again and ask for the list of available files (the same ones we sent to S3 in the previous workflow step)

peetucket commented on lib/robots/dor_repo/speech_to_text/stt_create.rb:
```diff
@@ -11,10 +11,51 @@ def initialize
+
+        def aws_provider
+          @aws_provider ||= Dor::TextExtraction::AwsProvider.new(region: Settings.aws.region, access_key_id: Settings.aws.access_key_id, secret_access_key: Settings.aws.secret_access_key)
+        end
```
i wonder if there is a nicer way to make this `aws_provider` available to any robot that needs it (currently two) without having to reinitialize like this each time (kind of like how we make the workflow_service client available to all robots)

peetucket replied:
Pushed a commit to extract this a new base robot class that we can subclass from.  Not sure how much of an improvement it is other than reducing some duplicated code

jmartin-sul commented:
> I wonder if instead of reaching out to the cocina object itself to find the list of SSTable files to add to the job, we should instead just look in the AWS bucket folder that should have been created in the previous workflow step and just add any files there. Could be convinced of either approach, i think they both have pros/cons but probably doesn't matter too much one way or the other.

good question, i could also go either way on that.  maybe post-standup discussion for the devs?

jmartin-sul commented on lib/robots/dor_repo/speech_to_text/stt_create.rb:
```diff
@@ -4,17 +4,54 @@ module Robots
+                                          message_body:
+                                        })
+
+          logger.info("Sent SQS message for druid #{druid} to queue #{aws_provider.sqs_todo_queue_url}")
```
might be nice to log job id also, while we're at it

jmartin-sul replied:
https://github.com/sul-dlss/common-accessioning/issues/1369

peetucket replied:
Done in #1370

jmartin-sul commented on lib/robots/dor_repo/speech_to_text/stt_create.rb:
```diff
@@ -11,10 +11,51 @@ def initialize
+              max_line_count: 80,
+              beam_size: 10
+            }
+          }
```
i think i'd vote for throwing it in config to start.  as a follow-on touch up.  vague memory of discussion about configuring whisper options (in an argo modal when the workflow is applied?).  if we do that, i could see something where the defaults come from the settings, we merge in overrides from user entry via workflow variable, and then put that combined hash into the options for the message to whisper.
